src/nonLinPMMH.cpp

```diff
@@ -0,0 +1,146 @@
+// -*- mode: C++; c-indent-level: 4; c-basic-offset: 4; indent-tabs-mode: nil; -*-
+//
+// nonLinPMMH.cpp: Example 3.1 of Andrieu et al. (2010). Implementing particle marginal
+// Metropolis-Hastings for a toy non-linear state space model previously described in
+// Gordon et al. (1993) and Kitagawa (1996).
+//
+// Copyright (C) 2017         Dirk Eddelbuettel, Adam Johansen and Leah South
+//
+// This file is part of RcppSMC.
+//
+// RcppSMC is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 2 of the License, or
+// (at your option) any later version.
+//
+// RcppSMC is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with RcppSMC.  If not, see <http://www.gnu.org/licenses/>.
+
+#include "nonLinPMMH.h"
+
+namespace nonLinPMMH {
+    const double a_prior = 0.01;
+	const double b_prior = 0.01;
+}
+
+using namespace std;
+using namespace nonLinPMMH;
+
+
+// nonLinPMMH_impl() function callable from R via Rcpp::
+// [[Rcpp::export]]
+Rcpp::DataFrame nonLinPMMH_impl(arma::vec data, unsigned long lNumber, unsigned long lMCMCits) { 	
+  
+  try {
+	arma::vec sigv(lMCMCits), sigw(lMCMCits);
+	long lIterates = data.n_rows;
+    y = data;
+	
+	arma::vec loglike = arma::zeros(lMCMCits);
+	arma::vec logprior = arma::zeros(lMCMCits);
+	
+	double dRand;
+	double loglike_prop;
+	double logprior_prop;
+	
+    //Initialise and run the sampler
+    smc::sampler<double> Sampler(lNumber, HistoryType::NONE);
+	theta_prop.sigv = 10.0;
+	theta_prop.sigw = 10.0;
+		
+	sigv(0) = theta_prop.sigv;
+	sigw(0) = theta_prop.sigw;
+	
+	// Getting a particle filtering estimate of the log likelihood.
+	smc::moveset<double> Moveset(fInitialise, fMove, NULL);
+    Sampler.SetResampleParams(ResampleType::MULTINOMIAL, 0.5);
+    Sampler.SetMoveSet(Moveset);
+    Sampler.Initialise();
+	Sampler.IterateUntil(lIterates-1);
+	loglike(0) = Sampler.GetLogNCPath();
+	
+	// Inverse gamma prior
+	logprior(0) = logPrior(theta_prop);
+    
+	double MH_ratio;
+	for (unsigned int i = 1; i<lMCMCits; i++){
+		// RW proposal for parameters
+		theta_prop.sigv = R::rnorm(sigv(i-1),0.15);
```

Comment:
The code looks fine and as we've said before working is far more important than fast, but one thought which might be worth bearing in mind for the future.

In this context, we know how many normal and uniform random variables we need in advance. From a speed perspective, it would make more sense to call R::rnorm only once or twice to obtain vectors of these RVs and similarly to call R::runif only once rather than once per iteration. I.e. To generate vectors of all of the RVs required in  advance outside of the main loop. There's a memory cost associated with it, but one which isn't enormous and it's best to minimize the number of calls back to R. It might be interesting to use rprof (or profvis) to compare the two approaches, actually.